registration/include/pcl/registration/impl/ndt_omp.hpp

```diff
@@ -0,0 +1,810 @@
+      converged_ = delta_norm == 0;
+      return;
+    }
+
+    delta /= delta_norm;
+    delta_norm = computeStepLengthMT (transform, delta, delta_norm, step_size_, transformation_epsilon_ / 2, score, score_gradient, hessian, output);
+    delta *= delta_norm;
+
+    // Convert delta into matrix form
+    convertTransform(delta, transformation_);
+
+    transform += delta;
+
+    // Update Visualizer (untested)
+    if (update_visualizer_)
+      update_visualizer_ (output, std::vector<int>(), *target_, std::vector<int>() );
+
+    const double cos_angle = 0.5 * transformation_.template block<3, 3>(0, 0).trace() - 1;
+    const double translation_sqr = transformation_.template block<3, 1>(0, 3).squaredNorm();
+
+    nr_iterations_++;
+
+    if (nr_iterations_ >= max_iterations_ ||
+        ((transformation_epsilon_ > 0 && translation_sqr <= transformation_epsilon_) && (transformation_rotation_epsilon_ > 0 && cos_angle >= transformation_rotation_epsilon_)) ||
+        ((transformation_epsilon_ <= 0)                                             && (transformation_rotation_epsilon_ > 0 && cos_angle >= transformation_rotation_epsilon_)) ||
+        ((transformation_epsilon_ > 0 && translation_sqr <= transformation_epsilon_) && (transformation_rotation_epsilon_ <= 0)))
+    {
+      converged_ = true;
+    }
+  }
+
+  // Store transformation probability.  The realtive differences within each scan registration are accurate
+  // but the normalization constants need to be modified for it to be globally accurate
+  trans_probability_ = score / static_cast<double> (input_->size ());
+}
+
+
+template<typename PointSource, typename PointTarget> double
+NormalDistributionsTransformOMP<PointSource, PointTarget>::computeDerivatives (Eigen::Matrix<double, 6, 1> &score_gradient,
+                                                                            Eigen::Matrix<double, 6, 6> &hessian,
+                                                                            const PointCloudSource &trans_cloud,
+                                                                            const Eigen::Matrix<double, 6, 1> &transform,
+                                                                            bool compute_hessian)
+{
+  std::vector<double> scores(input_->size());
+  std::vector<Eigen::Matrix<double, 6, 1>, Eigen::aligned_allocator<Eigen::Matrix<double, 6, 1>>> score_gradients(input_->size());
+  std::vector<Eigen::Matrix<double, 6, 6>, Eigen::aligned_allocator<Eigen::Matrix<double, 6, 6>>> hessians(input_->size());
+  for (std::size_t i = 0; i < input_->size(); i++) {
+    scores[i] = 0;
+    score_gradients[i].setZero();
+    hessians[i].setZero();
+  }
+
+  // Precompute Angular Derivatives (eq. 6.19 and 6.21)[Magnusson 2009]
+  computeAngleDerivatives (transform);
+
+  // Update gradient and hessian for each point, line 17 in Algorithm 2 [Magnusson 2009]
+  #pragma omp parallel for num_threads(num_threads_) schedule(guided, 8)
```

Comment:
Feels like only these 2 lines are the difference. Can you point to the new additions in core logic?

Reply:
Major new additions are follows:

Neighbor voxel search method
- Direct voxel accessing methods are added
https://github.com/koide3/pcl/blob/8b6a5aa079c76b939beacb161a81a54f712a1dc0/registration/include/pcl/registration/impl/ndt_omp.hpp#L204

Thread safety
- ```point_(jacobian|hessian)_``` are made per-thread local variables for thread safety
https://github.com/koide3/pcl/blob/8b6a5aa079c76b939beacb161a81a54f712a1dc0/registration/include/pcl/registration/impl/ndt_omp.hpp#L195
- Summing of some variables are modified in a thread-safe manner
https://github.com/koide3/pcl/blob/8b6a5aa079c76b939beacb161a81a54f712a1dc0/registration/include/pcl/registration/impl/ndt_omp.hpp#L175
https://github.com/koide3/pcl/blob/8b6a5aa079c76b939beacb161a81a54f712a1dc0/registration/include/pcl/registration/impl/ndt_omp.hpp#L241
https://github.com/koide3/pcl/blob/8b6a5aa079c76b939beacb161a81a54f712a1dc0/registration/include/pcl/registration/impl/ndt_omp.hpp#L250

SSE optimization
- The scalar type of ```angular_(jacobian|hessian)_``` is changed to float
https://github.com/koide3/pcl/blob/8b6a5aa079c76b939beacb161a81a54f712a1dc0/registration/include/pcl/registration/ndt_omp.h#L517
- A loop in the calculation of ```score_gradient``` is unrolled (this brings a large improvement on the single-thread registration speed, because a large portion of time is spent for this calculation)
https://github.com/koide3/pcl/blob/8b6a5aa079c76b939beacb161a81a54f712a1dc0/registration/include/pcl/registration/impl/ndt_omp.hpp#L394